src/Session.js

```diff
@@ -48,8 +34,20 @@ export default class Session {
         if (this.options.sessionToken && !requireNewToken) {
             return this.options.sessionToken
         }
-        const tokenObj = await this.loginFunction()
-        this.options.sessionToken = tokenObj.token
-        return tokenObj.token
+        if (this.state === Session.State.LOGGING_IN) {
+            return Promise.reject(new Error('Already logging in!'))
```

Comment:
Multiple `getSessionToken` calls would now automatically fail except the first one. This is a common situation when for example data produce requests occur in quick succession before the `loginFunction` completes. I think the functionality we want is this (this test case should be added):

- call `getSessionToken()` (returns `promise1`)
- assert `loginFunction` called once
- call `getSessionToken()` again (returns `promise2`)
- assert `loginFunction` still called once
- `loginFunction` resolves
- use `Promise.all` to combine `promise1` and `promise2`, then in `then` block check that both promises resolved with the **same** session token

Another test case (login failure):

- call `getSessionToken()` (returns `promise1`)
- assert `loginFunction` called once
- call `getSessionToken()` again (returns `promise2`)
- assert `loginFunction` still called once
- `loginFunction` **rejects**
- both promises should reject

Ok, so how to implement this behavior? How to make the second (and third, etc.) promise to resolve in a delayed fashion?

Two approaches come to mind:

**1. Return the original promise**

Save the Promise returned by `loginFunction` into some instance field and return that same Promise for every subsequent call if the state is `LOGGING_IN`. This would be an elegant solution but I'm not 100% sure if it works in practice.

**2. Wait for state change**

If we are `LOGGING_IN`, instead of the current auto-rejection, return a Promise that e.g. periodically checks if the state has changed to `LOGGED_IN`, and if it has, resolve the promise with the `sessionToken`. If the state has changed to `LOGGED_OUT` (you should add a `catch` block to the `loginFunction` call to set this state), reject the promise. Something like (completely untested):

```
if (this.state === Session.State.LOGGING_IN) {
    return new Promise((resolve, reject) => {
        let attempt = 0
        const check = () => {
            if (this.state === Session.State.LOGGED_IN) {
                resolve(this.options.sessionToken)
            } else if (this.state === Session.State.LOGGED_OUT || attempt > MAX) {
                reject(new Error('Was waiting for login, but it failed or timed out!'))
            } else {
                attempt += 1
                setTimeout(check, 100)
            }
        }
        setTimeout(check, 100)
    }
}
```

Reply:
My latest comments relate to this, so I'll comment here too: I'm not sure what you mean by "private function", as there are no actual private methods in JS? Anyway, I would propose a unit test where calls to outside of `Session` can easily be mocked, allowing you to implement the above outlined test cases.